docs/source/chunk.rst

```diff
@@ -0,0 +1,352 @@
+* Instantiate `ChunkDatasetOptions`
+* Instantiate `ChunkDataset`
+* Instantiate `ChunkDatasetWrapper`
+* Instantiate `DataLoader`
+* [optional] Set `DataLoader`'s :attr:`batch_size=None` to disable `auto collation`
+* [optional] Set `DataLoader`'s :attr:`pin_memory=True` to pin memory to GPU
+* [optional] If DataLoader's :attr:`num_workers` > 1:
+
+  * Implement worker initialization function
+  * Set :attr:`worker_init_fn` on :py:class:`DataLoader` constructor
+
+* Iterate over `DataLoader`
+
+**C++ snippet:**
+
+*Check `test/cpp/api/dataloader.h` and `test/cpp_extensions/extension.cpp` for full implementation*
+
+.. code-block:: cpp
+
+  #include <torch/extension.h>
+
+  using namespace torch::data::samplers;
+  using namespace torch::data::datasets;
+
+  /// Custom ChunkDataReader
+  class DummyChunkDataReader : public datasets::ChunkDataReader<int> {
+  public:
+    using BatchType = datasets::ChunkDataReader<int>::ChunkType;
+    using DataType = datasets::ChunkDataReader<int>::ExampleType;
+
+    /// Read an entire chunk.
+    BatchType read_chunk(size_t chunk_index) override {
+      BatchType batch_data;
+      int start_index = chunk_index == 0
+          ? 0
+          : std::accumulate(chunk_sizes, chunk_sizes + chunk_index, 0);
+
+      batch_data.resize(chunk_sizes[chunk_index]);
+
+      std::iota(batch_data.begin(), batch_data.end(), start_index);
+      return batch_data;
+    }
+
+    size_t chunk_count() override {
+      return chunk_count_;
+    };
+
+    void reset() override{};
+
+    const static size_t chunk_count_ = 3;
+    size_t chunk_sizes[chunk_count_] = {10, 5, 20};
+  };
+
+  PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
+    /// Specific DummyChunkDataReader implementation
+    auto dummy_reader = torch::data::datasets::bind_chunkdatareader<DummyChunkDataReader>(m, "DummyChunkDataReader");
+    dummy_reader.def(py::init<>(), "Create and return a new `DummyChunkDataReader` instance");
+
+    /// Specific DummyChunkDataset implementation
+    torch::data::datasets::bind_chunkdataset<DummyChunkDataReader>(m, "DummyChunkDataset");
+  }
+
+**Python snippet:**
+
+*Check `test/test_cpp_extensions.py (test_dummy_chunkdataset_bindings)` for full implementation*::
+
+>>> # Importing modules
+>>> from torch.utils.data import DataLoader
+>>> import torch.utils.data.chunk as chunk
+>>> import torch.utils.cpp_extension
+>>> # Dummy parameters
+>>> batch_size=5
+>>> cache_size=100
+>>> preloaders=1
+>>> # Main ChunkDataset classes
+>>> reader = cpp_extension.DummyChunkDataReader()
+>>> opt = chunk.ChunkDatasetOptions(preloader_count=preloaders, batch_size=batch_size, cache_size=cache_size)
+>>> dummy_chunkdataset = cpp_extension.DummyChunkDataset(chunk_reader=reader, options=opt, shuffle_chunks=False, shuffle_samples=False)
+>>> trainset = chunk.ChunkDataset(dummy_chunkdataset)
+>>> trainset.reset()
+>>> # Integrating with Python DataLoader
+>>> trainloader = DataLoader(dataset=trainset, num_workers=1)
```

Comment:
Did you really want to spawn a single new process? If you don't want to create any workers just don't specify the argument